src/main/java/com/endava/internship/collections/StudentMap.java

```diff
@@ -1,78 +1,248 @@
     }
 
     @Override
     public Integer get(Object o) {
-        //TODO
-        return null;
-    }
-
-    @Override
-    public Integer put(Student student, Integer integer) {
-        //TODO
-        return null;
+        if (!containsKey(o))
+            return null;
+        else
+            return getValueByKey(head, o);
     }
 
     @Override
     public Integer remove(Object o) {
-        //TODO
-        return null;
+        if (!findKey(head,o))
+            return null;
+        else {
+            return removeByKey(o);
+        }
     }
-
     @Override
     public void putAll(Map<? extends Student, ? extends Integer> map) {
-        //TODO
+        for (Entry<? extends Student, ? extends Integer> node : map.entrySet()){
+            Node newNode = new Node(node.getKey(), node.getValue());
+            head = insert(head, newNode);
+        }
     }
 
     @Override
     public void clear() {
-        //TODO
+        head = clearTree(head);
+        size = 0;
     }
 
     @Override
     public Set<Student> keySet() {
-        //TODO
-        return null;
+        print(head);
+        Set<Student> studentSet = new HashSet<>();
+        getKeySet(head, studentSet);
+        return studentSet;
     }
 
     @Override
     public Collection<Integer> values() {
-        //TODO
-        return null;
+        Collection<Integer> valueCollection = new ArrayList<>();
+        getAllValues(head, valueCollection);
+        return valueCollection;
     }
 
     @Override
     public Set<Entry<Student, Integer>> entrySet() {
         //Ignore this for homework
         throw new UnsupportedOperationException();
     }
+    private void print(Node current){
+        if (current != null){
+            print(current.getLeft());
+            System.out.println(current.getKey());
+            print(current.getRight());
+        }
+    }
+
+    private Node insert(Node root, Node newNode ){
+        if (root == null){
+            root = new Node(newNode.getKey(), newNode.getValue());
+            size++;
+            return root;
+        }
+
+        if (root.getKey().compareTo(newNode.getKey()) < 0){
+            root.setLeft(insert(root.getLeft(),newNode));
+        }
+        else if (root.getKey().compareTo(newNode.getKey()) > 0){
+            root.setRight(insert(root.getRight(), newNode));
+        }
+        else if (root.getKey().equals(newNode.getKey())){
+            root.setValue(newNode.getValue());
+        }
+        return root;
+    }
+    private boolean findKey(Node current,Object key){
+        if (current == null){
+            return false;
+        }
+        if (current.getKey().equals(key)){
+            return true;
+        }
+        if (current.getKey().compareTo((Student) key) > 0){
+            return findKey(current.getRight(), key);
+        }
+        else {
+            return findKey(current.getLeft(), key);
+        }
+    }
+    public boolean findValue(Node current, Object value, Boolean isPresent){
```

Comment:
make this method private. 
This method does not return a value if it finds it, it just tells if it exists or not so rename it to something like "isValuePresent"

Reply:
Fixed.